### PR TITLE
Add data-tags and data-title attributes to the ViewTemplate

### PR DESCRIPTION
**Pull request by pmario**
Functionality, new and modified Docs see: http://2737.tiddlyspot.com/#Modified%20Docs

The problem, that this PR solves: 

 - The existing core adds special classes to the tiddler body
 - Those classes are defined by tags.

The problem now is, that the html class attribute can't handle spaces and special chars. So the tag title must be url encoded. eg: ` tc-tagged-How%20to%20apply%20custom%20styles` for the title: `How to apply custom styles`

This is a problem for users. .. 

 - Users expect the possibility to use all UNICODE characters. 

This PR fixes this problem.

This PR replaces: PR #2737

### CODE REVIEW

**Comment by tobibeer** [comment]
To sum it up: For me, the problem is **not** that css classes might be less readable, at all. The problem is rather if I cannot address certain tagged tiddlers or certain titled titlers via css.

> The problem now is, that the html class attribute can't handle spaces and special chars. So the tag title must be url encoded

This is basically a false statement because this is an invalid css class:

`.tc-tagged-How%20to%20apply%20custom%20style`

As [said before](https://github.com/Jermolene/TiddlyWiki5/pull/2737#issuecomment-348758363), I would rather be in favor of fixing the broken implementation of **tc-tagged**.

Most importantly, your mechanism adds all tags to one property. This is a problem, not a solution, since we cannot logically combine tags then.

```
.tc-tagged-Foo_bar..tc-tagged-Mumble_frotz {
}
```

Therefore, I am not at all a friend of deprecating tc-tagged. I want to see it fixed.

In addition, I would prefer an implementation such as this alternatively to "data-tags":

http://apply.tiddlyspot.com

...since it adds even more benefits and flexibility. In short, it allows to apply ANY FILTER to create style classes and is not constrained to only tags or titles.

No matter what the implementation, I would also want to see some **tc-titled**.

**Comment by pmario** [comment]
>Most importantly, your mechanism adds all tags to one property. This is a problem, not a solution, since we cannot logically combine tags then.

That statement is simply wrong!

----------------

Here's how CSS works:

1) CSS classes are assigned from right to left. 

In the following code `bar` will overwrite rules for `foo`. That's hardcoded into browser behaviour. 

```
<span class="bar foo"></span>
```

This behaviour is a problem, since "tag order" is unpredictable. We want to remove this "undefined" behaviour. This PR does this. 


2) CSS definitions

Here `.foo` definition will overwrite `.bar` ... That's hardcoded into browser behaviour

```
.bar {
  display: block;
}

.foo {
  display: inline;
}
```

So depending on the settings and order from 1) AND 2) you will get different behaviour. 

-------

The following code works fine and doesn't contain any substituted characters, 
[Docs for Attribute Selectors: ](http://2737.tiddlyspot.com/#Attribute%20Selectors)

Either tag `bar` or tag `froz` create a red border. Case 1) from above doesn't matter anymore. 

```
[data-tags~="bar"],
[data-tags~="froz"] {
  border: 2px solid red;
}
```

If tiddler has both: `bar` and `froz`, you get an orange border

```
[data-tags*="bar"][data-tags*="froz"] {
  border: 2px solid orange;
}
```

 - `bar` creates a green border and 
 - `froz` creates a red border. 
 - If both tags are assigned, it depends on 2)
   - In our case the border will be `orange`
 - Nice and predictable for users.  
 - No encoding needed ... Tags can be UNICODE chars. 


```
[data-tags~="bar"] {
  border: 2px solid blue;
}

[data-tags~="froz"] {
  border: 2px solid red;
}
```

 - bar creates a blue border
 - froz creates a red border
 - if you swap positions, you get it the other way around.

No magic .... all pure CSS stuff. 

and so on: If you have a look at the possible [Attribute Selectors](https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors) you'll see what's possible.

**Comment by pmario** [comment]
If you need special handling for a tiddler with a field: `rank` and the value: `species` you can do this: 

title: myStyles
tags: $:/tags/Stylesheet 

```
<$list filter="[rank[species]]"> 
[data-title^="<$view field=title/>"] .tc-tiddler-body {
    column-count: 2;
}
</$list> 
```

And you are good to go.

**Comment by tobibeer** [comment]
> So depending on the settings and order from 1) AND 2) you will get different behaviour.

But that seems to be "just the nature of things". If you classify something as both a book and an audiobook, you have to decide at some point if, representationally, one kicks in over the other. Otherwise, css class possibly are not the (only) relevant aspect to consider, but rather things like conditional viewtemplates.

In other words, if you arrive at a point where "unpredictable tag order" prevents you from achieving something, then you're relying on the wrong tools for the job.

**Comment by tobibeer** [comment]
Also, you see, this is not a reliable solution to me...

```
;[attr*=value]
:Represents an element with an attribute name of attr and whose value contains at least one occurrence of string "value" as substring.
```

...because I might have other tags called "value xyz" that would also be adressed... and so your mechanism of identifying tagging attributes is not as reliable as it must be for complext titles. So, in a way, it is slightly less limited as tc-tagged is right now, but almost as unreliable in this regard.

I would very much prefer a reliable implementation, even if not as readable (within stylesheets)... unless commented properly.

**Comment by tobibeer** [comment]
Apparently, this won't work:

```
[attr*=[[value xyz]]]
```

Does it?

**Comment by pmario** [comment]
`[attr*="value xyz"]`

**Comment by tobibeer** [comment]
> `[attr*="value xyz"]` ... it also works with cyrrillic chars ;)

As I said, that is not helpful, since it false positive is "value xyzz" and also "value xyz a" and so forth.

**Comment by pmario** [comment]
Just use this .. `[attr*="[[value xyz]]"]` or even `[attr*="[value xyz]"]`, if you want to avoid `value xyzz`

If we want to avoid: `xyzz` or `xxyz` and the tag we want, is `xyz` we can use: `[attr~="xyz"]`

You can also say: `[attr*="[["]` ... special styling for tiddlers that are tagged with a: `[[tag with spaces]]` ... 

With the dynamic stylesheet that I did show at: https://github.com/Jermolene/TiddlyWiki5/pull/3038#issuecomment-349011301 ... it's possible to create any combination, that `<$list filter=""` allows. So I'm pretty sure, that we can cover 99% of user-cases with `[attr` settings.  ... and the rest can be done with `<lsit filter` stuff.

**Comment by tobibeer** [comment]
Ha...

> `[attr*="[[value xyz]]"]``

Ok, so this definitely allows for some more flexibility for tags with spaces.

Howewer for individual titles, we still have some issues with tags with spaces,

> If we want to avoid: `xyzz` or `xxyz` and the tag we want, is xyz we can use: `[attr~="xyz"]`

One false positive will be `[[abc xyz oh]]`.

----

I suppose, the main argument for not fixing **tc-tagged** is that the classes become harder to read. As for the title, we could easily add **tc-titled**, equivalent to **tc-tagged**.

**Comment by pmario** [comment]
>I suppose, the main argument for not fixing tc-tagged is that the classes become harder to read. As for the title, we could easily add tc-titled, equivalent to tc-tagged.

`tc-titled` ... That's the same problem. ... Users want to add the "exact same title" into the stylesheet. Everything else is confusing and hard to maintain.

**Comment by pmario** [comment]
>One false positive will be [[abc xyz oh]]

yea. ... If that's really a problem, CSS cascading will nicely deal with it. ... define a rule for `xyz` first and then one for `[[abc xyz oh]]` ... If that's a problem you can go with the `<$list filter` option.

**Comment by tobibeer** [comment]
> Users want to add the "exact same title" into the stylesheet. 

I beg to differ. I want to be able to address and style tiddler related elements. That's all. I do not care one bit if there was an underscore or some non-readable `fe23` where need be.

Truth is, I will not have to maintain complex stylesheets all the time, and if I do, I know what to look for.

>  If that's really a problem, CSS cascading will nicely deal with it.

Cascades hardly deal with that in a graceful manner. I surely don't want to take off another style from another thing to get my looks right some other place. Plus, it's highly confusing.

**Comment by twMat** [comment]
Note:

>CSS classes are assigned from right to left.
>In the following code bar will overwrite rules for foo. That's hardcoded into browser behaviour.
>
><span class="bar foo"></span>
>This behaviour is a problem, since "tag order" is unpredictable. We want to remove this "undefined" behaviour. This PR does this.

I believe this is incorrect; Writing "bar foo" or "foo bar" does not matter. Only the order in which the style definitions are evaluated in matter (or if they have different specificity).

**Comment by twMat** [comment]
>>I suppose, the main argument for not fixing tc-tagged is that the classes become harder to read. As for the title, we could easily add tc-titled, equivalent to tc-tagged.

>tc-titled ... That's the same problem. ... Users want to add the "exact same title" into the stylesheet. Everything else is confusing and hard to maintain.

Could stylesheets by default be dynamic TW5-type and `.tc-title` could be parsed(?) into url-encoded (or whatever code makes it valid)?

`.tc-title-[[My tiddler]]` could even be parsed as a link then. 

**Side note**: The idea with two types of stylesheets is confusing as it is. Given that stylesheets in TW are created in an unorthodox way to begin with, it would probably be OK to use dynamic TW5 as default (but permit advanced users to set it to static). This might overall open up for more active stylesheet manipulation via TW .

**Comment by pmario** [comment]
>Could stylesheets by default be dynamic TW5-type and .tc-title could be parsed(?) into url-encoded (or whatever code makes it valid)?

probably yes, but it should be the last option and not the norm.

**Comment by twMat** [comment]
Why?

**Comment by pmario** [comment]
>Side note: The idea with two types of stylesheets is confusing as it is. Given that stylesheets in TW are created in an unorthodox way to begin with, it would probably be OK to use dynamic TW5 as default (but permit advanced users to set it to static). This might overall open up for more active stylesheet manipulation via TW .

Yes. But stylesheets are recalculated everytime a tiddler changes in the store. It already adds several 100ms to the UI rendering time. The last time I did some measurements it was between 50-300ms. Prior to FF55 ... That makes it hard to track it down. ... The easisest way for me was to only calculate the Stylesheets at browser reload. .. I personally don't change stylesheets dynamically. Once I'm fine they stay constant for years to come. ... It's bloatware.

**Comment by tobibeer** [comment]
> .tc-title-[[My tiddler]] could even be parsed as a link then.

As Mario said, it would be possible, but it would have some performance impact. How much and when should be more thoroughly looked into, imho.

> stylesheets are recalculated everytime a tiddler changes in the store

Is that really so? Because, if it is, I would think it begs to be changed. Not sure how it could be achieved that they are only recalculated if a tiddler tagged **$:/tags/Stylesheet** changes. I guess the problem lies in using transclusion and macros in StyleSheets, which Mat wants to leverage. :-)

I can imagine a "refresh styles" button / action that triggers the refreshing, so TiddlyWiki is **not** constantly doing it, but only when one would reasonable expect it to (palette change) or ask it to (button click).

As for simplifying the ways to address a tiddler in a stylesheet, this would probably work...

```
<<.tc-title "foo bar">> {
}

<<.tc-tagged "foo bar">> {
}
```

And, if one wanted to remove the performance implications or use it in a stylesheet where parsing rules for macros are turned off, one could always replace either one with the parsed equivalent. Yes, it's mildly more complicated for a short moment, but then my whole point in all this is also...

> I personally don't change stylesheets dynamically. Once I'm fine they stay constant for years to come.

**Comment by pmario** [comment]
>Is that really so?

Yes. ... It has a "change" event handler. 

https://github.com/Jermolene/TiddlyWiki5/blob/9b2d9169464754a3fddfba0efd098724dbf649d1/core/modules/startup/render.js#L48

> Because, if it is, I would think it begs to be changed. Not sure how it could be achieved that they are only recalculated if a tiddler tagged $:/tags/Stylesheet changes. I guess the problem lies in using transclusion and macros in StyleSheets, which Mat wants to leverage. :-)

Right. ... Since $:/tags/Stylesheet tagged tiddlers can transclude content from other tiddlers, which are not tagged. ... I would call this **a dynamic StyleSheet**.

Just did some new performance tests. ... It has improved quite a bit. .. and I'm not sure anymor, if I didn't confuse the styleSheet numbers with "mainRefresh" numbers. ... But it is still in the 10+ ms. per UI change. eg: tiddler open, tiddler close, ... since those actioins change the $:/StoryList tiddler 

Variation seems to be much lower now too :)

**Comment by pmario** [comment]
Just to let you know. .... We are completely OT with the discussion atm.

**Comment by tobibeer** [comment]
I mean no discredit or offense to your work on this. However, I am in fact discussing wether this implementation is the way to move forward and what alternatives exist that are equally or more so worth pursuing.

In that sense we are quite on topic. As a consequence, I am inclined to make two more alternative PRs. 

One based on fixing and extend the tc-tagged mechanism [as proposed here](https://github.com/Jermolene/TiddlyWiki5/pull/2737#issuecomment-348758363) alongside those [tc-xxx macros](https://github.com/Jermolene/TiddlyWiki5/pull/3038#issuecomment-350461953) discussed above to make it easier to handle.

The other being based on the proposed **apply** filter mechanism referenced in [my first comment](https://github.com/Jermolene/TiddlyWiki5/pull/3038#issuecomment-348947774). There is absolutely nothing beyond this PR about it. Its simply a different way to provide a robust and extensible solution for "styling tiddlers".

I think it is important to evaluate thoroughly if either method is preferable to the `data-xxx` method of this PR. Once merged, we are stuck with either for quite some time to come and implementations will anchor around their capabilities and limitations.

**Comment by pmario** [comment]
I'm ok, if you create PRs with your suggestion. The problem, that I have with http://apply.tiddlyspot.com is, that imo it's way to complicated. ... But we can discuss this at your PR ;)

**Comment by tobibeer** [comment]
Sure. The ability to apply a css class to any defined filter set in one rule ...actually yields simplification, I would think. ;-)

File: core/ui/ViewTemplate.tid [hunk 1]
```diff
@@ -6,6 +6,6 @@ tc-tiddler-frame tc-tiddler-view-frame $(missingTiddlerClass)$ $(shadowTiddlerCl
 \define folded-state()
 $:/state/folded/$(currentTiddler)$
 \end
-<$set name="storyTiddler" value=<<currentTiddler>>><$set name="tiddlerInfoState" value=<<qualify "$:/state/popup/tiddler-info">>><$tiddler tiddler=<<currentTiddler>>><div class=<<frame-classes>>><$list filter="[all[shadows+tiddlers]tag[$:/tags/ViewTemplate]!has[draft.of]]" variable="listItem"><$transclude tiddler=<<listItem>>/></$list>
+<$set name="storyTiddler" value=<<currentTiddler>>><$set name="tiddlerInfoState" value=<<qualify "$:/state/popup/tiddler-info">>><$tiddler tiddler=<<currentTiddler>>><div data-title=<<currentTiddler>> data-tags={{!!tags}} class=<<frame-classes>>><$list filter="[all[shadows+tiddlers]tag[$:/tags/ViewTemplate]!has[draft.of]]" variable="listItem"><$transclude tiddler=<<listItem>>/></$list>
```
Comment (Jermolene) [inline core/ui/ViewTemplate.tid]: Hi @pmario `data-title=<<currentTiddler>>` isn't consistent with the `data-tiddler-title` that we now have in the core. Could you kindly refactor?

Comment (pmario) [reply core/ui/ViewTemplate.tid]: Yea. ... I missed that one, or I did forget about it :)) ... Will fix it

Comment (pmario) [reply core/ui/ViewTemplate.tid]: @Jermolene ... What do you think about my proposal to add a "class" field similar to the "color" field of tiddlers. .. See: https://groups.google.com/d/msg/tiddlywiki/nlXjxlHRLmw/vKIAiMdGBQAJ TLDR; section?

Comment (Jermolene) [reply core/ui/ViewTemplate.tid]: The "class” field is a great idea, go for it, if you've time, thanks!

**Comment by pmario** [comment]
Still a conflict. ... Close this PR again and create a new one. ...

**Pull request closed without merging**